java/org/apache/coyote/http11/Http11Processor.java

```diff
@@ -915,8 +915,26 @@ protected final void prepareResponse() throws IOException {
                 headers.addValue(Constants.CONNECTION).setString(
                         Constants.CLOSE);
             }
-        } else if (!http11 && !getErrorState().isError()) {
-            headers.addValue(Constants.CONNECTION).setString(Constants.KEEPALIVE);
+        } else if (!getErrorState().isError()) {
+            if (!http11) {
+                headers.addValue(Constants.CONNECTION).setString(Constants.KEEPALIVE);
+            }
+
+            boolean connectionKeepAlivePresent =
+                isConnectionToken(request.getMimeHeaders(), Constants.KEEPALIVE);
+
+            if (connectionKeepAlivePresent) {
+                int keepAliveTimeout = protocol.getKeepAliveTimeout();
+
+                if (keepAliveTimeout > 0) {
+                    String value = "timeout=" + keepAliveTimeout / 1000L;
+                    headers.setValue("Keep-Alive").setString(value);
+                }
+
+                if (http11) {
+                    headers.addValue(Constants.CONNECTION).setString(Constants.KEEPALIVE);
```

Comment:
Shouldn't this be inside the previous if block? I think it would be odd to send `Connection: keep-alive` on an HTTP 1.1 connection where there is no `Keep-Alive: ...` header.

Reply:
@markt-asf  This is is a very good question. I share @ChristopherSchultz's opinion. It is in line with https://tools.ietf.org/id/draft-thomson-hybi-http-timeout-01.html#rfc.section.2.

If the client has not requested a keep alive, the else clause is never evaluated. The condition is that the has send a `keep-alive` and either a timeout is set or the timeout is negative which means the connection can be held open forever. Does this make sense?

Reply:
Not really. I don't see the spec language that says you need to add `Connection: keep-alive` to an HTTP/1.1 connection when there is no `Keep-alive: ...` header.

Reply:
That might be true. I simply have implemented it just like in [HTTPd](https://github.com/apache/httpd/blob/trunk/modules/http/http_protocol.c#L283-L296). The header is written as soon as the client requests it. I wanted both to have the same behavior.

Reply:
I care more about what the spec says than what httpd does. While httpd is a useful guide, it may not follow the spec in all instances.

Reply:
Rereading the draft and the RFC 7230, I see not point which obligates us to respond with `Connection: Keep-Alive` on `HTTP/1.1` if the connection can be kept forever. So I have no counter arguments and will move and nest it.

Reply:
As far as I am concerned, it's a weird compatibility thing for HTTP/1.0 clients pretending to do HTTP/1.1.